src/compiler/scala/tools/nsc/backend/jvm/BTypesFromSymbols.scala

```diff
@@ -422,12 +422,12 @@ abstract class BTypesFromSymbols[G <: Global](val global: G) extends BTypes {
 
     val nestedClasses = {
       val ph = phase
-      Lazy(enteringPhase(ph)(nestedClassSymbolsNoJavaModuleClasses.map(classBTypeFromSymbol)))
+      Lazy.withLock(enteringPhase(ph)(nestedClassSymbolsNoJavaModuleClasses.map(classBTypeFromSymbol)))
```

Comment:
I wonder if we can identify certain `Lazy`s that we know will be forced, depending on the callsite where the `ClassBType` / `ClassInfo` is created from (code generation in particular)? Maybe you could experiment a bit (track from where the `ClassBType`s are created and see which ones will be completed)?

Reply:
I did have a brief try with making it more eager. I agree that if we know its going to be needed it better to do it more eagerly though.

I didn't spend much time on that though

Reply:
in my testing of the 3 call sites that use the lock about 70% of them were expanded.

Another thought that I had was that the background threads could expand the non symbol Lazy values more eagerly if we know some that we will need ( the withoutLock values). This doesn't occur without optimisation as I see it and that isn't my focus really

Reply:
@lrytz I went back and had another look. Some improvements added and I think that there are more to come.Not sure if they are part of this PR though.Will discuss via email